starknet_py/net/models/transaction.py

```diff
@@ -1,29 +1,238 @@
-from typing import Sequence, Union
+import base64
+import gzip
+import json
+import warnings
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from typing import Any, Dict, List, Optional, Sequence, TypeVar, Union
 
-from starkware.starknet.core.os.transaction_hash.transaction_hash import (
+import marshmallow
+import marshmallow_dataclass
+from marshmallow import fields
+from starkware.starknet.services.api.contract_class import ContractClass
+
+from starknet_py.cairo.selector import get_selector_from_name
+from starknet_py.net.client_models import TransactionType
+from starknet_py.net.models import compute_address
+from starknet_py.net.models.chains import StarknetChainId
+from starknet_py.net.schemas.common import Felt, NoneFelt
+from starknet_py.utils.crypto.transaction_hash import (
     TransactionHashPrefix,
-    calculate_transaction_hash_common,
+    compute_declare_transaction_hash,
+    compute_deploy_account_transaction_hash,
+    compute_transaction_hash,
 )
 
-# noinspection PyPep8Naming
-from starkware.starknet.definitions.transaction_type import TransactionType as TT
-from starkware.starknet.public.abi import get_selector_from_name
 
-# noinspection PyPep8Naming
-from starkware.starknet.services.api.gateway.transaction import Declare as DCL
-from starkware.starknet.services.api.gateway.transaction import DeployAccount as DAC
-from starkware.starknet.services.api.gateway.transaction import InvokeFunction as IF
-from starkware.starknet.services.api.gateway.transaction import Transaction as T
+@dataclass(frozen=True)
+class Transaction(ABC):
+    """
+    StarkNet transaction base class.
+    """
 
-from starknet_py.net.models.chains import StarknetChainId
-from starknet_py.utils.crypto.facade import pedersen_hash
-from starknet_py.utils.docs import as_our_module
+    version: int = field(metadata={"marshmallow_field": Felt()})
+
+    @property
+    @abstractmethod
+    def type(self) -> TransactionType:
+        """
+        Returns the corresponding TransactionType enum.
+        """
+
+    @abstractmethod
+    def calculate_hash(self, chain_id: StarknetChainId) -> int:
+        """
+        Calculates the transaction hash in the StarkNet network - a unique identifier of the
+        transaction. See :ref:`compute_transaction_hash` docstring for more details.
+        """
+
+
+@dataclass(frozen=True)
+class AccountTransaction(Transaction):
```

Comment:
I think (or actually my IDE thinks), this should also have `ABC` as base class, since it's abstract.